SPARK-5747: Fix wordsplitting bugs in make-distribution.sh

dyross opened the pull request:
The `$MVN` command variable may have spaces, so when referring to it, must wrap in quotes.

dyross committed:
SPARK-5747: Fix wordsplitting bugs in make-distribution.sh

Add double quotes to several variables, including $MVN.

AmplabJenkins commented:
Can one of the admins verify this patch?

nchammas commented:
I'm guessing this is a continuation of #4126.

dyross commented:
@nchammas it is related, but #4126 was a fix to a previous break. This change is a fix after this commit: https://github.com/apache/spark/commit/e902dc443dcc20f904e628b79b430b456cb330e4

Sorry, I should have specified that in the description.

nchammas commented:
@dyross Thank you for fixing this issue. Could you:
* Create a sub-task under [SPARK-5747](https://issues.apache.org/jira/browse/SPARK-5747) for `make-distribution.sh`
* Update the title of this PR to reference the new JIRA issue
* Resolve all other word splitting issues with `make-distribution.sh`

Totally fine if you just want to fix the call to Maven, but there are a few other problematic places in this script.

For example:
* [sub-shell output may contain spaces](https://github.com/apache/spark/blob/fa6bdc6e819f9338248b952ec578bcd791ddbf6d/make-distribution.sh#L101)
* [`$RPM_JAVA_HOME` may contain spaces](https://github.com/apache/spark/blob/fa6bdc6e819f9338248b952ec578bcd791ddbf6d/make-distribution.sh#L103)

nchammas commented:
FWIW, this patch LGTM as-is.

srowen commented:
Rather than add yet more JIRAs, let's just make this part of the SPARK-5747 umbrella. We may have two commits, sure. Putting the JIRA in the title, and perhaps addressing a few more instances of this in the script, would be great.

@pwendell quick check -- if we make a fix like this, how far back does it get back-ported? 1.2 or further?

dyross commented:
I've updated the commit wording and fixed a few more instances, as requested. The script seems to execute fine in our build system, but I am not 100% sure about all the edge cases.

nchammas commented on make-distribution.sh:
```diff
@@ -100,7 +100,7 @@ if [ -z "$JAVA_HOME" ]; then
   if [ $(command -v  rpm) ]; then
     RPM_JAVA_HOME=$(rpm -E %java_home 2>/dev/null)
```
I think we need to quote the subshell, since its output may contain spaces.

```
RPM_JAVA_HOME="$(rpm -E %java_home 2>/dev/null)"
```

dyross replied:
Done.

nchammas commented on make-distribution.sh:
```diff
@@ -147,7 +147,7 @@ if [[ ! "$JAVA_VERSION" =~ "1.6" && -z "$SKIP_JAVA_TEST" ]]; then
   echo "$JAVA_VERSION"
   read -p "Would you like to continue anyways? [y,n]: " -r
-  if [[ ! $REPLY =~ ^[Yy]$ ]]; then
+  if [[ ! "$REPLY" =~ ^[Yy]$ ]]; then
```
Since we are using the Bash-specific `[[ ... ]]` syntax here, there is actually no word splitting, so the original line should be safe. But I don't think it hurts to quote `$REPLY` anyway.

dyross replied:
Ok, I'll leave it quoted since it's consistent with other places. But thanks for the note.

nchammas commented:
Thanks for the updates @dyross! One minor comment and I think this is good to go.

nchammas commented:
This patch LGTM.

dyross commented:
Thanks for the review, @nchammas.

srowen commented:
ok to test

SparkQA commented:
[Test build #27381 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27381/consoleFull) for   PR 4540 at commit [`5a41596`](https://github.com/apache/spark/commit/5a41596d2cb1bc639c7786979cac2de7b23ffd64).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27381 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27381/consoleFull) for   PR 4540 at commit [`5a41596`](https://github.com/apache/spark/commit/5a41596d2cb1bc639c7786979cac2de7b23ffd64).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27381/
Test PASSed.

andrewor14 commented:
Merging into master 1.3 and 1.2 thanks
